website/utils/url_utils.ts

```diff
@@ -53,59 +63,114 @@ export type ViewerChannelSettingJson = {
+  /** Slice position per X, Y, and Z axes, as a list of comma-separated floats.
+   * 0.5 for all axes by default (e.g. `0.5,0.5,0.5`)
+   */
+  slice?: string = undefined;
+  /** Frame number, for time-series volumes. 0 by default. */
+  t?: string = undefined;
+}
+
+/** URL parameters that define data sources when loading volumes. */
+class DataParams {
   /**
    * One or more volume URLs to load. If multiple URLs are provided, they should
    * be separated by commas.
    */
-  url?: string;
+  url?: string = undefined;
   /**
    * The name of a dataset in the Cell Feature Explorer database. Used with `id`.
    */
-  dataset?: string;
+  dataset?: string = undefined;
   /**
    * The ID of a cell within the loaded dataset. Used with `dataset`.
    */
-  id?: string;
-  /**
-   * Axis to view. Valid values are "3D", "X", "Y", and "Z". Defaults to "3D".
-   */
-  view?: string;
-};
-// Copy of keys for above params. Both types are defined so that spec comments can be provided.
-// TODO: Remove redundant `baseParamKeys` type.
-const baseParamKeys = ["mask", "url", "dataset", "id", "view"] as const;
-const deprecatedParamKeys = ["ch", "luts", "colors"] as const;
+  id?: string = undefined;
+}
 
-type BaseParamKeys = (typeof baseParamKeys)[number];
-type ChannelKey = `c${number}`;
-type DeprecatedParamKeys = (typeof deprecatedParamKeys)[number];
-type AllParamKeys = BaseParamKeys | DeprecatedParamKeys | ChannelKey;
+class DeprecatedParams {
+  /** Deprecated query parameter for channel settings. */
+  ch?: string = undefined;
+  /** Deprecated query parameter for LUT settings. */
+  luts?: string = undefined;
+  /** Deprecated query parameter for channel colors. */
+  colors?: string = undefined;
+}
 
-type Params = BaseParams & { [_ in AllParamKeys]?: string };
+type Params = Partial<ViewerStateParams & DataParams & DeprecatedParams & ChannelParams>;
 
-const isParamKey = (key: string): key is BaseParamKeys => baseParamKeys.includes(key as BaseParamKeys);
-const isDeprecatedParamKey = (key: string): key is DeprecatedParamKeys =>
-  deprecatedParamKeys.includes(key as DeprecatedParamKeys);
-const isChannelKey = (key: string): key is ChannelKey => CHANNEL_STATE_KEY_REGEX.test(key);
+const allowedParamKeys: Array<keyof Params> = [
+  ...Object.keys(new ViewerStateParams()),
+  ...Object.keys(new DataParams()),
+  ...Object.keys(new DeprecatedParams()),
+] as Array<keyof Params>;
+const isParamKey = (key: string): key is keyof Params => allowedParamKeys.indexOf(key as keyof Params) !== -1;
+const isChannelKey = (key: string): key is keyof ChannelParams => CHANNEL_STATE_KEY_REGEX.test(key);
 
+/**
+ * Filters a set of URLSearchParams for only the keys that are valid parameters for the viewer.
+ * Non-matching keys are discarded.
+ * @param searchParams Input URL search parameters.
+ * @returns a dictionary object matching the type of `Params`.
+ */
 export function urlSearchParamsToParams(searchParams: URLSearchParams): Params {
```

Comment:
nit but this name confuses me. maybe `urlSearchParamsToAllowableParams` or `getAllowedParams` or `cleanUrlSearchParams`. I think `Params` as a type could also be changes to `AppParams` or `AllowableParams`